File: src/abc/abc.java
package abc;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.StringTokenizer;

class abc {
  private void run() throws IOException {
    BufferedReader r = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter w = new BufferedWriter(new OutputStreamWriter(System.out));

    StringTokenizer s = new StringTokenizer(r.readLine());
    int[] nums = { Integer.parseInt(s.nextToken()), Integer.parseInt(s.nextToken()), Integer.parseInt(s.nextToken()) };
    Arrays.sort(nums);

    String line = r.readLine();
    r.close();

    switch (line) {
      case "ABC":
        w.write(nums[0] + " " + nums[1] + " " + nums[2] + "\n");
        break;
      case "ACB":
        w.write(nums[0] + " " + nums[2] + " " + nums[1] + "\n");
        break;
      case "BAC":
        w.write(nums[1] + " " + nums[0] + " " + nums[2] + "\n");
        break;
      case "BCA":
        w.write(nums[1] + " " + nums[2] + " " + nums[0] + "\n");
        break;
      case "CAB":
        w.write(nums[2] + " " + nums[0] + " " + nums[1] + "\n");
        break;
      case "CBA":
        w.write(nums[2] + " " + nums[1] + " " + nums[0] + "\n");
        break;
    }

    w.flush();
  }

  public static void main(String[] a) throws IOException {
    (new abc()).run();
  }

}

File: src/abc/abc.hs
import Data.List (intercalate, sort)
import Data.List.NonEmpty (NonEmpty ((:|)), (<|))
import qualified Data.List.NonEmpty as NE

main :: IO ()
main = interact (formatOutput . solve . parseInput)

-- helpers for parsing
splitOn' :: Char -> String -> NonEmpty String
splitOn' sep s = case s of { [] -> "" :| []; (part : remainder) -> if part == sep then "" <| splitOn' sep remainder else let head :| tail = splitOn' sep remainder in (part : head) :| tail}

splitOn :: Char -> String -> [String]
splitOn sep s = NE.toList (splitOn' sep s)

-- Define input and output here
type Input = ([Int], String)

type Output = (Int, Int, Int)

formatOutput :: Output -> String
formatOutput (a, b, c) = show a ++ " " ++ show b ++ " " ++ show c ++ "\n"

parseInput :: String -> Input
parseInput rawInput =
  let
    -- map read (splitOn ',' rawInput)
    l    = lines rawInput
    nums = sort (map read (splitOn ' ' (l !! 0)))
    str  = l !! 1
  in
    (nums, str)

--parseTestcases :: String -> [Input]
--parseTestcases rawInput = map parseInput tail (lines rawInput)

solve :: Input -> Output
solve (nums, str) =
  let
    a = nums !! 0
    b = nums !! 1
    c = nums !! 2
  in case str of
    "ABC" -> (a, b, c)
    "ACB" -> (a, c, b)
    "BAC" -> (b, a, c)
    "BCA" -> (b, c, a)
    "CAB" -> (c, a, b)
    "CBA" -> (c, b, a)
    _     -> error "should not happen"
